lib/service.js

```diff
@@ -78,93 +76,111 @@ function processService(serviceMethod, env) {
+                    throw error;
+                }
+            })
             .then(() => responseBody);
+    }
 
-        function checkStatusCode(response) {
-            //eslint-disable-next-line eqeqeq
-            if (response.statusCode != '200') {
-
-                debug('Unable to reach API :(');
+    function onSuccess(data) {
+        return sendResponse({ status: 'success', data });
+    }
 
-                var error = new Error(util.format(
-                    'Failed to POST data to %s (%s, %s)',
-                    POST_RESULT_URL, response.statusCode, response.body
-                ));
-                error.sendable = false;
-                throw error;
-            }
+    function onError(err) {
+        if (err.sendable === false) {
+            throw new Error(err.message);
         }
+        const errorData = {
+            message: err.message,
+        };
+        log.error(err, err.stack);
+        return sendResponse({ status: 'error', data: errorData });
     }
 
+    return Q.fcall(init)
+        .spread(execService)
+        .then(onSuccess)
+        .catch(onError);
+
     function verifyCredentials(cfg, params) {
+        const callScope = new EventEmitter();
+        callScope.logger = new ComponentLogger();
+
         function doVerification(verify) {
-            return new Promise((resolve, reject) => {
-                function legacyCallback(e, result) {
+            // eslint-disable-next-line no-async-promise-executor
+            return new Promise(async (resolve, reject) => {
+                async function legacyCallback(e, result) {
                     if (e) {
-                        return reject(e);
+                        reject(e);
                     }
                     resolve(result);
                 }
-                var callScope = new EventEmitter();
-                callScope.logger = new ComponentLogger();
-                const result = verify.call(callScope, cfg, legacyCallback);
 
-                if (result) {
-                    resolve(result);
+                try {
+                    let result = verify.call(callScope, cfg, legacyCallback);
+                    /**
+                     * Runs verify with an await statement if and only if it requires one, depending on if the existing
+                     * result statement has resolved or not. BUGFIX to handle both cases component4 and component12,
+                     * based on https://stackoverflow.com/a/35820220
+                     * @param {Promise} p
+                     */
+                    async function resolvePromise(p) {
+                        const t = {};
+                        return Promise.race([p, t])
+                            .then(async (v) => {
+                                if (v === t) result = await verify.call(callScope, cfg, legacyCallback);
+                            });
+                    }
+                    await resolvePromise(result);
```

Comment:
This ended up being a bit complicated to solve, because await on something that is not an async function was having conflicting effects. This is a solution that doesn't require checking if a function is async to do its checks, and instead relies on whether there is an unfinished promise.

This was required because if a `verify` returns `Promise.resolve()` to result that is awaited, the value returned still needs to be resolved, even though it is `undefined` (see test case component4). This conditionally resolves the result correctly.